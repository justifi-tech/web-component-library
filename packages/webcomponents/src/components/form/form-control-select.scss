@import 'reboot';
@import 'forms';

:host {
  @include host-base-styles;
  display: block;
}

select[disabled] {
  background-color: #e9ecef;
  opacity: 0.5;
  cursor: not-allowed;
}
