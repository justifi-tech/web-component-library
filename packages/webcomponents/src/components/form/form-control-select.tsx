import {
  Component,
  Host,
  h,
  Prop,
  Event,
  EventEmitter,
  Watch,
} from '@stencil/core';

@Component({
  tag: 'form-control-select'
})
export class SelectInput {
  selectElement!: HTMLSelectElement;

  @Prop() label: string;
  @Prop() name: any;
  @Prop() helpText?: string;
  @Prop() errorText?: string;
  @Prop() defaultValue: string;
  @Prop() inputHandler: (name: string, value: string) => void;
  @Prop() options: { label: string; value: string }[];
  @Prop() disabled: boolean;

  @Event() formControlInput: EventEmitter<any>;
  @Event() formControlBlur: EventEmitter<any>;

  @Watch('defaultValue')
  handleDefaultValueChange(newValue: string) {
    this.updateInput(newValue);
  }

  updateInput(newValue: any) {
    this.selectElement.value = newValue;
  }

  handleFormControlInput(event: any) {
    const target = event.target;
    const name = target.getAttribute('name');
    this.inputHandler(name, target.value);
    this.formControlInput.emit(target.value);
  }

  componentDidLoad() {
    this.updateInput(this.defaultValue);
  }

  render() {
    return (
      <Host exportparts="label,input,input-invalid">
        <div class="form-group d-flex flex-column">
          <label part="label" class="form-label" htmlFor={this.name}>
            {this.label}
          </label>
          <select
            ref={el => (this.selectElement = el as HTMLSelectElement)}
            id={this.name}
            name={this.name}
            onInput={(event: any) => this.handleFormControlInput(event)}
            onBlur={() => this.formControlBlur.emit()}
            part={`input ${this.errorText ? 'input-invalid' : ''}`}
            class={this.errorText ? 'form-select is-invalid' : 'form-select'}
            disabled={this.disabled}
          >
            {this.options?.map(option => (
              <option value={option.value}>{option.label}</option>
            ))}
          </select>
          <form-control-help-text helpText={this.helpText} />
          <form-control-error-text errorText={this.errorText} />
        </div>
      </Host>
    );
  }
}
