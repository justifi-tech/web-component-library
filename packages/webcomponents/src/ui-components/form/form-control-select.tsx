import {
  Component,
  Host,
  h,
  Prop,
  Event,
  EventEmitter,
  Watch,
  State,
} from '@stencil/core';
import { FormControlErrorText } from '../../ui-components';
import { input, inputDisabled, inputFocused, inputInvalid, label } from '../../styles/parts';

@Component({
  tag: 'form-control-select'
})
export class SelectInput {
  selectElement!: HTMLSelectElement;

  @State() isFocused: boolean = false;

  @Prop() name: any;
  @Prop() label: string;
  @Prop() options: { label: string; value: string }[];
  @Prop() inputHandler: (name: string, value: string) => void;
  @Prop() defaultValue?: string;
  @Prop() helpText?: string;
  @Prop() errorText?: string;
  @Prop() disabled?: boolean;

  @Watch('defaultValue')
  handleDefaultValueChange(newValue: string) {
    this.updateInput(newValue);
  }

  @Event() formControlInput: EventEmitter<any>;
  @Event() formControlBlur: EventEmitter<any>;

  componentDidLoad() {
    this.updateInput(this.defaultValue);
  }

  handleFormControlInput = (event: any) => {
    const target = event.target;
    const name = target.getAttribute('name');
    this.inputHandler(name, target.value);
    this.formControlInput.emit(target.value);
  }

  updateInput = (newValue: any) => {
    this.selectElement.value = newValue;
  }

  private get part() {
    let part = input;
    if (this.errorText) {
      part = inputInvalid;
    }
    if (this.disabled) {
      part = inputDisabled;
    }
    if (this.isFocused) {
      part = inputFocused;
    }
    return part;
  }

  render() {
    return (
      <Host>
        <div class="form-group d-flex flex-column">
          <div class="d-flex align-items-start gap-2">
            <label part={label} class="form-label" htmlFor={this.name}>
              {this.label}
            </label>
            <form-control-tooltip helpText={this.helpText} />
          </div>
          <select
            ref={el => (this.selectElement = el as HTMLSelectElement)}
            id={this.name}
            name={this.name}
            onFocus={() => this.isFocused = true}
            onBlur={() => {
              this.isFocused = false;
              this.formControlBlur.emit();
            }}
            onInput={this.handleFormControlInput}
            part={this.part}
            class={this.errorText ? 'form-select is-invalid' : 'form-select'}
            disabled={this.disabled}
          >
            {this.options?.map(option => (
              <option value={option.value}>{option.label}</option>
            ))}
          </select>
          <FormControlErrorText errorText={this.errorText} name={this.name} />
        </div>
      </Host>
    );
  }
}
