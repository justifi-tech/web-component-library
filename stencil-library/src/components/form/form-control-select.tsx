import {
  Component,
  Host,
  h,
  Prop,
  Event,
  EventEmitter,
  Watch,
} from '@stencil/core';

@Component({
  tag: 'form-control-select',
  styleUrl: 'form-control-select.scss',
  shadow: true,
})
export class SelectInput {
  selectElement!: HTMLSelectElement;

  @Prop() label: string;
  @Prop() name: any;
  @Prop() error: string;
  @Prop() defaultValue: string;
  @Prop() inputHandler: (name: string, value: string) => void;
  @Prop() options: { label: string; value: string }[];
  @Prop() disabled: boolean;
  @Event() formControlInput: EventEmitter<any>;
  @Event() formControlBlur: EventEmitter<any>;

  @Watch('defaultValue')
  handleDefaultValueChange(newValue: string) {
    this.updateInput(newValue);
  }

  updateInput(newValue: any) {
    this.selectElement.value = newValue;
  }

  handleFormControlInput(event: any) {
    const target = event.target;
    const name = target.getAttribute('name');
    this.inputHandler(name, target.value);
  }

  componentDidLoad() {
    this.updateInput(this.defaultValue);
  }

  render() {
    return (
      <Host exportparts="label,input,input-invalid">
        <label part="label" class="form-label" htmlFor={this.name}>
          {this.label}
        </label>
        <select
          ref={el => (this.selectElement = el as HTMLSelectElement)}
          id={this.name}
          name={this.name}
          onInput={(event: any) => this.handleFormControlInput(event)}
          onBlur={() => this.formControlBlur.emit()}
          part={`input ${this.error && 'input-invalid'}`}
          class={this.error ? 'form-select is-invalid' : 'form-select'}
          disabled={this.disabled}
        >
          {this.options.map(option => (
            <option value={option.value}>{option.label}</option>
          ))}
        </select>
        {this.error && <div class="invalid-feedback">{this.error}</div>}
      </Host>
    );
  }
}
