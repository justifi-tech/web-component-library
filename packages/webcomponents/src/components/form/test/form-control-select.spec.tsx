import { newSpecPage } from '@stencil/core/testing';
import { SelectInput } from '../form-control-select';

describe('form-control-select', () => {
  it('renders with default props', async () => {
    const page = await newSpecPage({
      components: [SelectInput],
      html: `<form-control-select label="Test Select" name="test"></form-control-select>`,
    });

    expect(page.root).toMatchSnapshot();
    expect(page.rootInstance.label).toBe('Test Select');
    expect(page.rootInstance.name).toBe('test');
  });

  it('populates options correctly and handles selection', async () => {
    const options = [
      { label: 'Option 1', value: '1' },
      { label: 'Option 2', value: '2' }
    ];
    const page = await newSpecPage({
      components: [SelectInput],
      html: `<form-control-select></form-control-select>`,
    });
    page.rootInstance.options = options;
    await page.waitForChanges();

    const selectElement = page.root.shadowRoot.querySelector('select');
    expect(selectElement.children.length).toBe(options.length);
    expect(selectElement.children[0].textContent).toBe(options[0].label);

    selectElement.value = options[1].value;
    expect(selectElement.value).toBe('2');
  });

  it('shows error and applies error styling when error prop is provided', async () => {
    const page = await newSpecPage({
      components: [SelectInput],
      html: `<form-control-select error="This field is required."></form-control-select>`,
    });

    const shadowRoot = page.root.shadowRoot;
    expect(shadowRoot.querySelector('.invalid-feedback').textContent).toBe('This field is required.');
    expect(shadowRoot.querySelector('.form-select').classList.contains('is-invalid')).toBeTruthy();
  });

  it('emits formControlInput event on input', async () => {
    const page = await newSpecPage({
      components: [SelectInput],
      html: `<form-control-select></form-control-select>`,
    });

    page.rootInstance.inputHandler = jest.fn();

    const inputEventSpy = jest.fn();
    page.root.addEventListener('formControlInput', inputEventSpy);

    const selectElement = page.root.shadowRoot.querySelector('select');
    selectElement.value = '1';
    selectElement.dispatchEvent(new Event('input', { bubbles: true, composed: true }));

    await page.waitForChanges();
    expect(inputEventSpy).toHaveBeenCalled();
  });

  it('emits formControlBlur event on blur', async () => {
    const page = await newSpecPage({
      components: [SelectInput],
      html: `<form-control-select></form-control-select>`,
    });

    const blurEventSpy = jest.fn();
    page.root.addEventListener('formControlBlur', blurEventSpy);

    const selectElement = page.root.shadowRoot.querySelector('select');
    selectElement.dispatchEvent(new Event('blur'));

    expect(blurEventSpy).toHaveBeenCalled();
  });

  it('disables select when disabled prop is true', async () => {
    const page = await newSpecPage({
      components: [SelectInput],
      html: `<form-control-select disabled="true"></form-control-select>`,
    });

    const selectElement = page.root.shadowRoot.querySelector('select');

    // This is weird, but the value of the disabled 
    // attribute is an empty string on Stencil mock DOM
    expect(selectElement.attributes.getNamedItem('disabled').value).toBe('');
  });

  it('handles empty options array', async () => {
    const page = await newSpecPage({
      components: [SelectInput],
      html: `<form-control-select></form-control-select>`,
    });
    page.rootInstance.options = [];
    await page.waitForChanges();

    const selectElement = page.root.shadowRoot.querySelector('select');
    expect(selectElement.children.length).toBe(0);
  });
});
