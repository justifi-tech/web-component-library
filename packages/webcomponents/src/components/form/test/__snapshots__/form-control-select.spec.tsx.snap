// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form-control-select Renders with all props provided 1`] = `
<form-control-select exportparts="label,input,input-invalid">
  <div class="d-flex flex-column form-group">
    <label class="form-label" htmlfor="delivery_method" part="label">
      Select your delivery method
    </label>
    <select class="form-select is-invalid" disabled="" id="delivery_method" name="delivery_method" part="input input-invalid">
      <option value="pickup">
        Pickup
      </option>
      <option value="delivery">
        Delivery
      </option>
    </select>
    <form-control-help-text>
      <small class="form-text text-muted" id="form-help-text-undefined">
        Select your preferred delivery method
      </small>
    </form-control-help-text>
    <form-control-error-text>
      <small class="form-text text-danger" id="form-error-text-undefined">
        This field is required.
      </small>
    </form-control-error-text>
  </div>
</form-control-select>
`;

exports[`form-control-select Renders with default props 1`] = `
<form-control-select exportparts="label,input,input-invalid">
  <div class="d-flex flex-column form-group">
    <label class="form-label" htmlfor="delivery_method" part="label">
      Select your delivery method
    </label>
    <select class="form-select" id="delivery_method" name="delivery_method" part="input "></select>
    <form-control-help-text></form-control-help-text>
    <form-control-error-text></form-control-error-text>
  </div>
</form-control-select>
`;
